File: continuous/distributions/beta_prime_4p.py
import scipy.optimize
import scipy.stats
import numpy
import scipy.special as sc

class BETA_PRIME_4P:
    """
    scale Prime 4p Distribution
    https://en.wikipedia.org/wiki/Beta_prime_distribution
    """
    def __init__(self, measurements):
        self.parameters = self.get_parameters(measurements)
        self.alpha = self.parameters["alpha"]
        self.beta = self.parameters["beta"]
        self.scale = self.parameters["scale"]
        self.loc = self.parameters["loc"]
        
    def cdf(self, x: float) -> float:
        """
        Cumulative distribution function
        Calculated using the definition of the function
        Alternative: quadrature integration method
        """
        z = lambda t: (t - self.loc) / self.scale
        # result = scipy.stats.betaprime.cdf(x, self.alpha, self.beta, loc=self.loc, scale=self.scale)
        result = sc.betainc(self.alpha, self.beta, z(x) / (1 + z(x)))
        return result
    
    def pdf(self, x: float) -> float:
        """
        Probability density function
        Calculated using definition of the function in the documentation
        """
        z = lambda t: (t - self.loc) / self.scale
        # result = scipy.stats.betaprime.pdf(x, self.alpha, self.beta, loc=self.loc, scale=self.scale)
        result = (1 / self.scale) * (z(x) ** (self.alpha - 1) * (1 + z(x)) ** (-self.alpha - self.beta)) / (sc.beta(self.alpha,self.beta))
        return result

    def get_num_parameters(self) -> int:
        """
        Number of parameters of the distribution
        """
        return len(self.parameters)
    
    def parameter_restrictions(self) -> bool:
        """
        Check parameters restrictions
        """
        v1 = self.alpha > 0
        v2 = self.beta > 0
        v3 = self.scale > 0
        return v1 and v2 and v3
    
    def get_parameters(self, measurements) -> dict[str, float | int]:
        """
        Calculate proper parameters of the distribution from sample measurements.
        The parameters are calculated by solving the equations of the measures expected 
        for this distribution.The number of equations to consider is equal to the number 
        of parameters.
        
        Parameters
        ==========
        measurements: MEASUREMESTS
            attributes: mean, std, variance, skewness, kurtosis, median, mode, min, max, length, num_bins, data

        Returns
        =======
        parameters : dict
            {"alpha":  * , "scale":  * , "min":  * , "max":  * }
        """
        ## In this distribution solve the system is best than scipy estimation.
        def equations(initial_solution: tuple[float], measurements) -> tuple[float]:
            α, β, scale, loc = initial_solution
        
            parametric_mean = scale * α / (β - 1) + loc
            parametric_variance = (scale ** 2) * α * (α + β - 1) / ((β - 1) ** 2 * (β - 2))
            # parametric_skewness = 2 * math.sqrt(((β - 2)) / (α * (α + β - 1))) * (((2 * α + β - 1)) / (β - 3))
            parametric_median = loc + scale * scipy.stats.beta.ppf(0.5, α, β) / (1 - scipy.stats.beta.ppf(0.5, α, β))
            parametric_mode = scale * (α - 1) / (β + 1) + loc
            
            eq1 = parametric_mean - measurements.mean
            eq2 = parametric_variance - measurements.variance
            # eq3 = parametric_skewness - measurements.skewness
            eq3 = parametric_median - measurements.median
            eq4 = parametric_mode - measurements.mode
            
            return (eq1, eq2, eq3, eq4)
        
        scipy_params = scipy.stats.betaprime.fit(measurements.data)
        
        try:
            bnds = ((0, 0, 0,  - numpy.inf), (numpy.inf, numpy.inf, numpy.inf, numpy.inf))
            x0 = (measurements.mean, measurements.mean, scipy_params[3], measurements.mean)
            args = ([measurements])
            solution = scipy.optimize.least_squares(equations, x0, bounds = bnds, args=args)
            parameters = {"alpha": solution.x[0], "beta": solution.x[1], "scale": solution.x[2], "loc": solution.x[3]}
        except:
            parameters = {"alpha": scipy_params[0], "beta": scipy_params[1], "scale": scipy_params[3], "loc": scipy_params[2]}
      
        return parameters

if __name__ == "__main__":
    ## Import function to get measurements
    import sys
    sys.path.append("../measurements")
    from measurements import MEASUREMENTS_CONTINUOUS

    ## Import function to get measurements
    def get_data(direction):
        sample_distribution_file = open(direction, "r")
        data = [float(x.replace(",", ".")) for x in sample_distribution_file.read().splitlines()]
        return data
    
    ## Distribution class
    path = "../data/data_beta_prime_4p.txt"
    data = get_data(path) 
    measurements = MEASUREMENTS_CONTINUOUS(data)
    distribution = BETA_PRIME_4P(measurements)
    
    print(distribution.get_parameters(measurements))
    print(distribution.cdf(measurements.mean))
    print(distribution.pdf(measurements.mean))

File: script-files/measures/outputsContinuous/beta_prime_4p.js
jStat = require("../node_modules/jstat");

dists = {
    beta_prime_4p: {
        measurements: {
            nonCentralMoments: function (k, alpha, beta, loc, scale) {
                return jStat.gammafn(k + alpha) * jStat.gammafn(beta - k) / (jStat.gammafn(alpha) * jStat.gammafn(beta))
            },
            centralMoments: function (k, alpha, beta, loc, scale) {
                const µ1 = this.nonCentralMoments(1, alpha, beta, loc, scale);
                const µ2 = this.nonCentralMoments(2, alpha, beta, loc, scale);
                const µ3 = this.nonCentralMoments(3, alpha, beta, loc, scale);
                const µ4 = this.nonCentralMoments(4, alpha, beta, loc, scale);

                let result;
                switch (k) {
                    case 1: result = 0; break;
                    case 2: result = µ2 - µ1 ** 2; break;
                    case 3: result = µ3 - 3 * µ1 * µ2 + 2 * µ1 ** 3; break;
                    case 4: result = µ4 - 4 * µ1 * µ3 + 6 * (µ1 ** 2) * µ2 - 3 * (µ1 ** 4); break;
                };
                return result
            },
            stats: {
                mean: function (alpha, beta, loc, scale) {
                    const µ1 = dists.beta_prime_4p.measurements.nonCentralMoments(1, alpha, beta, loc, scale);
                    return loc + scale * µ1;
                },
                variance: function (alpha, beta, loc, scale) {
                    const µ1 = dists.beta_prime_4p.measurements.nonCentralMoments(1, alpha, beta, loc, scale);
                    const µ2 = dists.beta_prime_4p.measurements.nonCentralMoments(2, alpha, beta, loc, scale);
                    return (scale ** 2) * (µ2 - µ1 ** 2);
                },
                standardDeviation: function (alpha, beta, loc, scale) {
                    return Math.sqrt(this.variance(alpha, beta, loc, scale));
                },
                skewness: function (alpha, beta, loc, scale) {
                    const µ1 = dists.beta_prime_4p.measurements.nonCentralMoments(1, alpha, beta, loc, scale);
                    const µ2 = dists.beta_prime_4p.measurements.nonCentralMoments(2, alpha, beta, loc, scale);
                    const std = Math.sqrt(µ2 - µ1 ** 2);
                    const central_µ3 = dists.beta_prime_4p.measurements.centralMoments(3, alpha, beta, loc, scale);
                    return central_µ3 / (std ** 3);
                },
                kurtosis: function (alpha, beta, loc, scale) {
                    const µ1 = dists.beta_prime_4p.measurements.nonCentralMoments(1, alpha, beta, loc, scale);
                    const µ2 = dists.beta_prime_4p.measurements.nonCentralMoments(2, alpha, beta, loc, scale);
                    const std = Math.sqrt(µ2 - µ1 ** 2);
                    const central_µ4 = dists.beta_prime_4p.measurements.centralMoments(4, alpha, beta, loc, scale);
                    return central_µ4 / (std ** 4);
                },
                median: function (alpha, beta, loc, scale) {
                    return dists.beta_prime_4p.measurements.ppf(0.5, alpha, beta, loc, scale);
                },
                mode: function (alpha, beta, loc, scale) {
                    return loc + scale * (alpha - 1) / (beta + 1);
                },
            },
        }
    }
}
console.log(dists.beta_prime_4p.measurements.stats.mean(100, 54, 0, 20))
console.log(dists.beta_prime_4p.measurements.stats.variance(100, 54, 0, 20))
console.log(dists.beta_prime_4p.measurements.stats.standardDeviation(100, 54, 0, 20))
console.log(dists.beta_prime_4p.measurements.stats.skewness(100, 54, 0, 20))
console.log(dists.beta_prime_4p.measurements.stats.kurtosis(100, 54, 0, 20))
// console.log(dists.beta_prime_4p.measurements.stats.median(100, 54, 0, 20))
console.log(dists.beta_prime_4p.measurements.stats.mode(100, 54, 0, 20))

// mean_value: 37.73584905660374
// variance_value: 41.898293945287435
// standard_deviation_value: 6.472889149775966
// skewness_value: 0.578410671752133
// kurtosis_value: 3.6318751249615646
// median_value: 37.14265555487496
// mode_value: 36